### ISSUE TITLE
Performance doubt about svd

### ISSUE THREAD

**Issue by SunYanCN** [mars-project/mars#247]
My code is here.

```
from time import time

def test_numpy():
    start_time = time()
    import numpy as np
    a = np.random.rand(5000, 4000)
    U, s, V = np.linalg.svd(a)
    print(time()-start_time)

def test_mars():
    import mars.tensor as mt
    start_time = time()
    a = mt.random.rand(5000, 4000,chunk_size=100)
    U, s, V = mt.linalg.svd(a).execute()
    print(time()-start_time)

if __name__ == '__main__':
    test_numpy()
    test_mars()
```
The result is:
```
36.3145694732666
141.55802178382874
```
Does this mean that this function is four or five times slower than numpy when the Tall Skinny QR (TSQR) Matrix Factorization condition is not met?

**Comment by NJUPole** [comment]
We refer to the method in the paper http://arxiv.org/abs/1301.1071 .

When input matrix is nearly a square matrix, there's no advantage for this method. For tall-and-skinny matrix, it will be split into multiple sub-matrices horizontally, so we will do `rechunk` in tiles even `chunk_size` is specified. In your case, the true chunk size  is 4000 to make sure split won't happen along the 2nd axis(row).

**Comment by NJUPole** [comment]
Actually, the performance is much better than Numpy when the matrix is tall and skinny. It's my early test:

```Python
In [1]: import numpy as np

In [2]: a = np.random.rand(20000, 100)

In [3]: %time U, s, V = np.linalg.svd(a)
CPU times: user 4min 3s, sys: 10.2 s, total: 4min 13s
Wall time: 1min 18s
```
``` Python
In [1]: import mars.tensor as mt

In [2]: a = mt.random.rand(20000, 100, chunk_size=100)

In [3]: %time U, s, V = mt.linalg.svd(a).execute()
CPU times: user 5.42 s, sys: 1.49 s, total: 6.91 s
Wall time: 1.87 s
```

**Comment by chineking** [comment]
> Does this mean that this function is four or five times slower than numpy when the Tall Skinny QR (TSQR) Matrix Factorization condition is not met?

Yes, you are right, when the TSQR algo is not suitable, the performance may be poor than numpy itself.

Super welcome to bring up other algorithms for the parallel SVD/QR/Cholesky/LU decomposition.

**Comment by SunYanCN** [comment]
I know that a random svd decomposition method can be parallelized and very fast, refer to the following two links:
[sklearn randomized_svd](https://scikit-learn.org/stable/modules/generated/sklearn.utils.extmath.randomized_svd.html#sklearn.utils.extmath.randomized_svd)
[Randomized SVD 算法介绍与实现](https://cloud.tencent.com/developer/article/1005571)

**Comment by chineking** [comment]
> I know that a random svd decomposition method can be parallelized and very fast, refer to the following two links:
> [sklearn randomized_svd](https://scikit-learn.org/stable/modules/generated/sklearn.utils.extmath.randomized_svd.html#sklearn.utils.extmath.randomized_svd)
> [Randomized SVD 算法介绍与实现](https://cloud.tencent.com/developer/article/1005571)

We shall do some investigation. If you have any interest in such work, let us know. Welcome to participate the development.

**Comment by SunYanCN** [comment]
I am from Huazhong University of Science and Technology. The job of the lab is to do tensor decomposition for the recommendation system. Recently I tried to use Mars as a backend for a tensor decomposition library to improve performance. I will try some improvements.

**Comment by chineking** [comment]
> I am from Huazhong University of Science and Technology. The job of the lab is to do tensor decomposition for the recommendation system. Recently I tried to use Mars as a backend for a tensor decomposition library to improve performance. I will try some improvements.

Great, your background and your work match well with Mars. 

I have noticed that you have proposed a few issues, some of them do give useful suggestions for Mars. Still I encourage you to step forward, and try to get involved in the part of Mars under the hood, it's hard indeed, thus I think it's helpful for both of your projects and your personal experience. It's all up to you though, suggestions about Mars already mean a lot for us.

Thanks all the same.

**Comment by chineking** [comment]
We are implementing `randomized_svd` in #388 , but the fancy indexing should be supported first.  I will update the progress in this thread. 

@SunYanCN FYI if you are interested.

**Comment by SunYanCN** [comment]
@chineking As a graduate student graduates next year, I am now busy preparing for work. However, I have recently seen the Alibaba Summer of Code, I intend to participate. Will this project participate? If this project is involved, it may be the best of both worlds for me.

**Comment by chineking** [comment]
> https://developer.aliyun.com/summerofcode2019

Sounds good, I will see if we can make Mars a project hosted.

**Comment by sighingnow** [comment]
I think the benchmark and conclusion in [issuecomment-468306739](https://github.com/mars-project/mars/issues/247#issuecomment-468306739) may not be very well to demonstrate the problem. You may need to run the `np.linalg.svd` with `full_matrices=False`, which has a notable performance impact, especially for tall and skinny matrices.

And in fact, mars does use the `full_matrices=False` setting when invoking `xp.linalg.svd`.

**Comment by chineking** [comment]
> I think the benchmark and conclusion in [issuecomment-468306739](https://github.com/mars-project/mars/issues/247#issuecomment-468306739) may not be very well to demonstrate the problem. You may need to run the `np.linalg.svd` with `full_matrices=False`, which has a notable performance impact, especially for tall and skinny matrices.

You are right,  `full_matrices` impact huge on performance, we also found that.

> And in fact, mars does use the `full_matrices=False` setting when invoking `xp.linalg.svd`.

Dig into the implementation of TSQR, given a matrix with shape (m, n) (M >> n) who has `k` chunk on the 1st axis, TSQR will  apply QR  to all chunks, and then we only do a SVD on matrix with shape (kn, n). I doubt `full_matrices` will not effect so much for `xp.linalg.svd` in Mars, because the matrix is such small.  Maybe compare TSQR with `full_matrices=False` SVD is not fair enough.

Anyway, TSQR can already solve quit a lot problem since the input matrix could be really massive.

**Comment by chineking** [comment]
@SunYanCN I have to inform you that I have contacted the organizers of summer of code, but we have missed the window for the projects part, so we cannot make our project hosted this year.

I'd like to close this issue anyway.
